File: iz-backend/pkg/signal/ratchet.go
package signal

import (
	"crypto/ecdh"
	"encoding/json"
	"errors"
	"fmt"
	"sync"
)

const (
	maxSkip      = 1000
	ratchetInfo  = "iz_ratchet_v1"
)

// RatchetHeader is included in every encrypted message so the recipient
// can advance their ratchet state.
type RatchetHeader struct {
	DHPublicKey string `json:"dh"`  // sender's current ratchet public key (base64)
	PrevCounter int    `json:"pn"`  // previous sending chain length
	Counter     int    `json:"n"`   // current message number
}

// EncryptedMessage is a Double Ratchet encrypted message ready to send over the wire.
type EncryptedMessage struct {
	Header     RatchetHeader `json:"header"`
	Ciphertext []byte        `json:"ciphertext"` // AES-256-GCM; header bytes used as AAD
}

// RatchetState holds the full Double Ratchet session state for one conversation.
// It must be serialised and stored (encrypted) by the client between sessions.
type RatchetState struct {
	mu sync.Mutex

	// Ratchet key pair for the sending direction
	DHSendPriv *ecdh.PrivateKey
	DHRecvPub  *ecdh.PublicKey

	RootKey      []byte
	SendChainKey []byte
	RecvChainKey []byte

	SendCount     int
	RecvCount     int
	PrevSendCount int

	// Skipped message keys: "b64pubkey:counter" -> msgKey
	SkippedKeys map[string][]byte
}

// InitSender initialises a RatchetState for the party that sends the first message (Alice).
// sharedKey comes from X3DH; bobSPKPub is Bob's signed prekey public key.
func InitSender(sharedKey []byte, bobSPKPub *ecdh.PublicKey) (*RatchetState, error) {
	// Generate Alice's initial sending ratchet key pair
	sendPriv, _, err := GenerateKeyPair()
	if err != nil {
		return nil, err
	}

	// Perform one ratchet step to derive the first sending chain
	dh, err := DH(sendPriv, bobSPKPub)
	if err != nil {
		return nil, err
	}
	rootKey, sendCK, err := HKDF(append(sharedKey, dh...), ratchetInfo)
	if err != nil {
		return nil, err
	}

	return &RatchetState{
		DHSendPriv:  sendPriv,
		DHRecvPub:   bobSPKPub,
		RootKey:     rootKey,
		SendChainKey: sendCK,
		SkippedKeys:  make(map[string][]byte),
	}, nil
}

// InitReceiver initialises a RatchetState for the party that receives the first message (Bob).
// sharedKey comes from X3DH; bobSPKPriv is Bob's signed prekey private key.
func InitReceiver(sharedKey []byte, bobSPKPriv *ecdh.PrivateKey, aliceRatchetPub *ecdh.PublicKey) (*RatchetState, error) {
	// Perform one receive ratchet step
	dh, err := DH(bobSPKPriv, aliceRatchetPub)
	if err != nil {
		return nil, err
	}
	rootKey, recvCK, err := HKDF(append(sharedKey, dh...), ratchetInfo)
	if err != nil {
		return nil, err
	}

	// Bob generates his own sending ratchet key pair
	sendPriv, _, err := GenerateKeyPair()
	if err != nil {
		return nil, err
	}
	dh2, err := DH(sendPriv, aliceRatchetPub)
	if err != nil {
		return nil, err
	}
	rootKey2, sendCK, err := HKDF(append(rootKey, dh2...), ratchetInfo)
	if err != nil {
		return nil, err
	}

	return &RatchetState{
		DHSendPriv:  sendPriv,
		DHRecvPub:   aliceRatchetPub,
		RootKey:     rootKey2,
		SendChainKey: sendCK,
		RecvChainKey: recvCK,
		SkippedKeys:  make(map[string][]byte),
	}, nil
}

// RatchetEncrypt encrypts a plaintext message, advancing the sending chain.
func (s *RatchetState) RatchetEncrypt(plaintext []byte) (*EncryptedMessage, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	nextCK, msgKey := KDFChain(s.SendChainKey)
	s.SendChainKey = nextCK

	header := RatchetHeader{
		DHPublicKey: EncodePublicKey(s.DHSendPriv.PublicKey()),
		PrevCounter: s.PrevSendCount,
		Counter:     s.SendCount,
	}
	s.SendCount++

	headerBytes, err := json.Marshal(header)
	if err != nil {
		return nil, err
	}
	ct, err := Encrypt(msgKey, plaintext, headerBytes)
	if err != nil {
		return nil, err
	}
	return &EncryptedMessage{Header: header, Ciphertext: ct}, nil
}

// RatchetDecrypt decrypts an incoming message, performing a DH ratchet step if necessary.
func (s *RatchetState) RatchetDecrypt(msg *EncryptedMessage) ([]byte, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	headerBytes, err := json.Marshal(msg.Header)
	if err != nil {
		return nil, err
	}

	// Check skipped message keys first
	if pt := s.trySkipped(msg.Header, headerBytes, msg.Ciphertext); pt != nil {
		return pt, nil
	}

	dhPub, err := DecodePublicKey(msg.Header.DHPublicKey)
	if err != nil {
		return nil, err
	}

	// Do we need a DH ratchet step?
	currentPub := ""
	if s.DHRecvPub != nil {
		currentPub = EncodePublicKey(s.DHRecvPub)
	}
	if currentPub != msg.Header.DHPublicKey {
		if err := s.skipKeys(msg.Header.PrevCounter); err != nil {
			return nil, err
		}
		if err := s.dhRatchet(dhPub); err != nil {
			return nil, err
		}
	}

	if err := s.skipKeys(msg.Header.Counter); err != nil {
		return nil, err
	}

	nextCK, msgKey := KDFChain(s.RecvChainKey)
	s.RecvChainKey = nextCK
	s.RecvCount++

	return Decrypt(msgKey, msg.Ciphertext, headerBytes)
}

// dhRatchet performs one DH ratchet step (receiving side).
func (s *RatchetState) dhRatchet(newDHPub *ecdh.PublicKey) error {
	s.PrevSendCount = s.SendCount
	s.SendCount = 0
	s.RecvCount = 0
	s.DHRecvPub = newDHPub

	dh, err := DH(s.DHSendPriv, newDHPub)
	if err != nil {
		return err
	}
	rootKey, recvCK, err := HKDF(append(s.RootKey, dh...), ratchetInfo)
	if err != nil {
		return err
	}
	s.RootKey = rootKey
	s.RecvChainKey = recvCK

	// Generate new sending ratchet key pair
	newSendPriv, _, err := GenerateKeyPair()
	if err != nil {
		return err
	}
	s.DHSendPriv = newSendPriv

	dh2, err := DH(newSendPriv, newDHPub)
	if err != nil {
		return err
	}
	rootKey2, sendCK, err := HKDF(append(s.RootKey, dh2...), ratchetInfo)
	if err != nil {
		return err
	}
	s.RootKey = rootKey2
	s.SendChainKey = sendCK
	return nil
}

// skipKeys stores skipped message keys up to the given counter.
func (s *RatchetState) skipKeys(until int) error {
	if s.RecvCount+maxSkip < until {
		return errors.New("too many skipped messages")
	}
	for s.RecvCount < until {
		nextCK, msgKey := KDFChain(s.RecvChainKey)
		s.RecvChainKey = nextCK
		k := fmt.Sprintf("%s:%d", EncodePublicKey(s.DHRecvPub), s.RecvCount)
		s.SkippedKeys[k] = msgKey
		s.RecvCount++
	}
	return nil
}

// trySkipped checks the skipped key map for the message key.
func (s *RatchetState) trySkipped(h RatchetHeader, headerBytes, ct []byte) []byte {
	k := fmt.Sprintf("%s:%d", h.DHPublicKey, h.Counter)
	msgKey, ok := s.SkippedKeys[k]
	if !ok {
		return nil
	}
	delete(s.SkippedKeys, k)
	pt, err := Decrypt(msgKey, ct, headerBytes)
	if err != nil {
		return nil
	}
	return pt
}

File: iz-web/src/lib/crypto/ratchet.ts
import { sharedSecret, generateKeyPair, KeyPair, toHex } from './x25519';
import { deriveKey } from './hkdf';
import { encrypt, decrypt } from './aes-gcm';

export interface RatchetState {
  rootKey: Uint8Array;
  sendChainKey: Uint8Array | null;
  recvChainKey: Uint8Array | null;
  remoteRatchetPub: Uint8Array | null;
  localRatchetKeyPair: KeyPair;
  sendCount: number;
  recvCount: number;
  prevSendCount: number;
  skippedMessageKeys: Record<string, Uint8Array>; // { "pub_hex:index": key }
}

const MAX_SKIP = 1000;

/**
 * Initializes a new ratchet state for a conversation.
 */
export function initRatchet(rootKey: Uint8Array, remotePub: Uint8Array | null, isInitiator: boolean): RatchetState {
  const state: RatchetState = {
    rootKey,
    sendChainKey: null,
    recvChainKey: null,
    remoteRatchetPub: remotePub,
    localRatchetKeyPair: generateKeyPair(),
    sendCount: 0,
    recvCount: 0,
    prevSendCount: 0,
    skippedMessageKeys: {},
  };

  if (isInitiator && remotePub) {
    const { rootKey: newRoot, chainKey } = kdfRoot(state.rootKey, sharedSecret(state.localRatchetKeyPair.privateKey, remotePub));
    state.rootKey = newRoot;
    state.sendChainKey = chainKey;
  }

  return state;
}

/**
 * Encrypts a message and advances the sending chain.
 */
export async function ratchetEncrypt(state: RatchetState, plaintext: string) {
  if (!state.sendChainKey) {
    throw new Error('Ratchet not initialized for sending');
  }

  const { chainKey: newChain, messageKey } = kdfChain(state.sendChainKey);
  state.sendChainKey = newChain;

  const { ciphertext, iv } = await encrypt(plaintext, messageKey);
  
  const header = {
    ratchetPub: state.localRatchetKeyPair.publicKey,
    prevSendCount: state.prevSendCount,
    index: state.sendCount,
  };

  state.sendCount++;

  return {
    header,
    ciphertext,
    iv,
  };
}

/**
 * Decrypts a message and advances the receiving chain.
 */
export async function ratchetDecrypt(
  state: RatchetState,
  ciphertext: Uint8Array,
  iv: Uint8Array,
  header: { ratchetPub: Uint8Array; prevSendCount: number; index: number }
): Promise<string> {
  // 1. Check if message key was skipped and stored
  const keyTag = `${toHex(header.ratchetPub)}:${header.index}`;
  if (state.skippedMessageKeys[keyTag]) {
    const mk = state.skippedMessageKeys[keyTag];
    delete state.skippedMessageKeys[keyTag];
    return decrypt(ciphertext, mk, iv);
  }

  // 2. DH Ratchet if remote public key changed
  if (!state.remoteRatchetPub || !compareUint8Arrays(header.ratchetPub, state.remoteRatchetPub)) {
    skipMessageKeys(state, header.prevSendCount);
    dhRatchet(state, header.ratchetPub);
  }

  // 3. Skip message keys if index is ahead
  skipMessageKeys(state, header.index);

  // 4. Advance receiving chain
  if (!state.recvChainKey) throw new Error('No receiving chain');
  const { chainKey: newChain, messageKey } = kdfChain(state.recvChainKey);
  state.recvChainKey = newChain;
  state.recvCount++;

  return decrypt(ciphertext, messageKey, iv);
}

function dhRatchet(state: RatchetState, remotePub: Uint8Array) {
  state.prevSendCount = state.sendCount;
  state.sendCount = 0;
  state.recvCount = 0;
  state.remoteRatchetPub = remotePub;

  // Advance root key with new DH secret
  const { rootKey: r1, chainKey: c1 } = kdfRoot(state.rootKey, sharedSecret(state.localRatchetKeyPair.privateKey, remotePub));
  state.rootKey = r1;
  state.recvChainKey = c1;

  // Generate new local ratchet key pair
  state.localRatchetKeyPair = generateKeyPair();

  // Advance root key again for sending chain
  const { rootKey: r2, chainKey: c2 } = kdfRoot(state.rootKey, sharedSecret(state.localRatchetKeyPair.privateKey, remotePub));
  state.rootKey = r2;
  state.sendChainKey = c2;
}

function skipMessageKeys(state: RatchetState, untilIndex: number) {
  if (state.recvCount + MAX_SKIP < untilIndex) {
    throw new Error('Too many skipped messages');
  }
  if (state.recvChainKey) {
    while (state.recvCount < untilIndex) {
      const { chainKey: newChain, messageKey } = kdfChain(state.recvChainKey);
      state.recvChainKey = newChain;
      const keyTag = `${toHex(state.remoteRatchetPub!)}:${state.recvCount}`;
      state.skippedMessageKeys[keyTag] = messageKey;
      state.recvCount++;
    }
  }
}

function compareUint8Arrays(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}

// ─── Internal KDFs ─────────────────────────────────────────────────────────────

function kdfRoot(rootKey: Uint8Array, dhSecret: Uint8Array) {
  const output = deriveKey(dhSecret, rootKey, 'WhisperRatchet', 64);
  return {
    rootKey: output.slice(0, 32),
    chainKey: output.slice(32),
  };
}

function kdfChain(chainKey: Uint8Array) {
  const output = deriveKey(new Uint8Array([0x01]), chainKey, 'WhisperMessageKeys', 32);
  const nextChainKey = deriveKey(new Uint8Array([0x02]), chainKey, 'WhisperMessageKeys', 32);
  return {
    chainKey: nextChainKey,
    messageKey: output,
  };
}
